redaxo/src/addons/phpmailer/lib/mailer.php

```diff
@@ -348,4 +357,158 @@ public static function errorMail(): void
+            $tokenData = [
+                'client_id' => $this->graphClientId,
+                'scope' => 'https://graph.microsoft.com/.default',
+                'client_secret' => $this->graphClientSecret,
+                'grant_type' => 'client_credentials',
+            ];
+
+            try {
+                $tokenResponse = $tokenSocket->doPost($tokenData);
+                $token = json_decode($tokenResponse->getBody(), true);
+                $token['expires'] = time() + ($token['expires_in'] ?? 3600);
+
+                $accessToken = $token['access_token'] ?? null;
+                if (!$accessToken) {
+                    throw new Exception(rex_i18n::msg('phpmailer_msgraph_no_token'));
+                }
+
+                rex_config::set('phpmailer', 'msgraph_token', $token);
+            } catch (Exception $e) {
+                $this->setError(rex_i18n::msg('phpmailer_msgraph_auth_error') . $e->getMessage());
+                $accessToken = null;
+                rex_config::set('phpmailer', 'msgraph_token', null);
+            }
+        }
+
+        // Mail senden via rex_socket
+        $mailUrl = "https://graph.microsoft.com/v1.0/users/$from/sendMail";
+        $mailSocket = rex_socket::factoryUrl($mailUrl);
+        $mailSocket->addHeader('Authorization', 'Bearer ' . $accessToken);
+        $mailSocket->addHeader('Content-Type', 'application/json');
+        $mailData = [
+            'message' => [
+                'subject' => $subject,
+                'body' => $body,
+                'toRecipients' => $to,
+                'from' => ['emailAddress' => ['address' => $from]],
+            ],
+            'saveToSentItems' => true,
+        ];
+        if (!empty($cc)) {
+            $mailData['message']['ccRecipients'] = $cc;
+        }
+        if (!empty($bcc)) {
+            $mailData['message']['bccRecipients'] = $bcc;
+        }
+        if (is_array($replyTo) && count($replyTo) > 0) {
+            $mailData['message']['replyTo'] = $replyTo;
+        }
+        if (!empty($attachments)) {
+            $mailData['message']['attachments'] = $attachments;
+        }
+        if (!empty($customHeaders)) {
+            $mailData['message']['internetMessageHeaders'] = $customHeaders;
+        }
+        if ('' !== $this->ConfirmReadingTo) {
+            // MS Graph API unterstützt keine Read-Receipts an beliebige Empfänger
+            $mailData['message']['isReadReceiptRequested'] = true;
+        }
+
+        if (rex::isDebugMode()) {
+            // Debug: JSON-Body loggen ins REDAXO-Addon-Data-Verzeichnis
+            $debugPath = rex_path::addonData('phpmailer', 'graph_mail_debug.json');
+            rex_file::put($debugPath, json_encode($mailData, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
+        }
+        try {
+            $mailResponse = $mailSocket->doPost(json_encode($mailData));
+            if ($mailResponse->getStatusCode() >= 400) {
```

Comment:
```suggestion
            if (!$mailResponse->isSuccessful()) {
```

Sollte, warum auch immer, ein 3xx-Fehler kommen, sollte es auch als Fehler behandelt werden. Werden aktuell wie erfolgreich behandelt, aber der Umleitung wird nicht gefolgt. Wobei Umleitung bei POST-Requests auch ungewöhnlich wären (inzwischen aber mit Status 307/308 ja möglich).